MANTA-4762 Want integration tests for manatee-primary-resolver

isaacdavis opened the pull request:
See the README for instructions on running the test suite. Here's how it looks in my build zone:

```
[isaac@buildzone ~/git-sand/rust-cueball/resolvers/manatee-primary-resolver]$ RESOLVER_LOG_LEVEL=critical cargo test | bunyan
    Finished dev [unoptimized + debuginfo] target(s) in 0.18s
     Running /home/isaac/git-sand/rust-cueball/target/debug/deps/cueball_manatee_primary_resolver-51a225306b48e151

running 13 tests
test test::process_value_test_invalid_json ... ok
test test::process_value_test_no_port_pg_url ... ok
test test::process_value_test_no_pg_url ... ok
test test::process_value_test_no_ip ... ok
test test::process_value_test_ip_change ... ok
test test::process_value_test_port_ip_change ... ok
test test::process_value_test_wrong_type_pg_url ... ok
test test::process_value_test_invalid_pg_url ... ok
test test::process_value_test_no_change ... ok
test test::process_value_test_wrong_type_ip ... ok
test test::process_value_test_port_change ... ok
test test::process_value_test_invalid_ip ... ok
test test::prop_zk_connect_string_parse ... ok
     Running /home/isaac/git-sand/rust-cueball/target/debug/deps/echo_resolver-5ae3b9b85b915625

test result: ok. 13 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running /home/isaac/git-sand/rust-cueball/target/debug/deps/channel_test-d71dd0ec29a7cc7b

running 2 tests
test channel_test_start_with_closed_rx ... ok
test channel_test_exit_upon_closed_rx ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running /home/isaac/git-sand/rust-cueball/target/debug/deps/connection_test-01603380863668b4

running 2 tests
test connection_test_reconnect_after_zk_hiccup ... ok
test connection_test_start_with_unreachable_zookeeper ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running /home/isaac/git-sand/rust-cueball/target/debug/deps/watch_test-3c11d62963df703f

running 16 tests
test watch_test_valid_to_no_pg_url ... ok
test watch_test_valid_to_invalid_ip ... ok
test watch_test_port_change ... ok
test watch_test_port_ip_change ... ok
test watch_test_invalid_to_valid ... ok
test watch_test_no_change ... ok
test watch_test_valid_to_no_port_pg_url ... ok
test watch_test_valid_to_invalid_pg_url ... ok
test watch_test_invalid_to_invalid ... ok
test watch_test_ip_change ... ok
test watch_test_valid_to_wrong_type_ip ... ok
test watch_test_valid_to_invalid_json ... ok
test watch_test_valid_to_wrong_type_pg_url ... ok
test watch_test_valid_to_no_ip ... ok
test watch_test_nonexistent_node ... ok
test watch_test_disappearing_node ... ok

test result: ok. 16 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

   Doc-tests cueball-manatee-primary-resolver

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

kellymclaughlin commented:
I think what we should do to keep the CI checks happy is to move the tests that require zookeeper under a `tests` subdirectory rather than `src`. That way `cargo` will deem them integration tests and we can update the Makefile target to only run the unit tests for CI like we did [here](https://github.com/joyent/manta-buckets-mdapi/blob/master/Makefile#L100-L102) with `buckets-mdapi`.

isaacdavis commented:
The integration tests are already under a `tests` subdirectory -- it's just their `common` code that's under `src`. This is the only way I could find to avoid a circular dependency between the `common` code (needed in the unit tests in `lib.rs`) and the `manatee-primary-resolver` module itself. There might be a more idiomatic way?

Anyway -- I added a new Makefile target and made Jenkins use it. I also bumped the minor version number, because in the course of writing the integration tests I found and fixed a bug.

kellymclaughlin commented on resolvers/manatee-primary-resolver/src/bin/echo-resolver.rs:
```diff
@@ -0,0 +1,130 @@
+//
```
I have a couple of suggestions for a perhaps better place to put this. We could make it an example program which I think would work since it interacts with the manatee-primary-resolver via its standard public API. Another option is to make it a separate crate (perhaps under `tools` at the top of the workspace?). With either option we get to keep the manatee-primary-resolver as a simple library crate and we can avoid having to bring in some of the extra build dependencies that are only need by the echo-resolver program. What do you think?

isaacdavis replied:
This is a good idea! I went with the latter suggestion and gave it its own crate under a new `tools` directory at the top of the workspace.

isaacdavis commented on resolvers/manatee-primary-resolver/src/bin/echo-resolver.rs:
```diff
@@ -0,0 +1,130 @@
+//
+// Copyright 2019 Joyent, Inc.
```
done

isaacdavis commented on resolvers/manatee-primary-resolver/tests/connection_test.rs:
```diff
@@ -0,0 +1,130 @@
+//
+// Copyright 2019 Joyent, Inc
```
done

isaacdavis commented on resolvers/manatee-primary-resolver/tests/watch_test.rs:
```diff
@@ -0,0 +1,277 @@
+//
+// Copyright 2019 Joyent, Inc
```
done

isaacdavis commented on resolvers/manatee-primary-resolver/tests/channel_test.rs:
```diff
@@ -0,0 +1,123 @@
+//
+// Copyright 2019 Joyent, Inc
```
done

isaacdavis commented on resolvers/manatee-primary-resolver/Cargo.toml:
```diff
@@ -1,34 +1,32 @@
 serde_json = "1.0.40"
 slog = { version = "2.4.1", features = [ "max_level_trace" ] }
-slog-stdlog = "3"
+slog-bunyan = { git = "https://github.com/slog-rs/bunyan" }
```
done
